File: packages/runpane/src/daemonClient.ts
import { createHash } from 'node:crypto';
import net from 'node:net';
import os from 'node:os';
import path from 'node:path';
import { StringDecoder } from 'node:string_decoder';
import { boundary, decodeBoundary } from './boundaryDecoder';
import type { BoundarySchema, JsonValue } from './boundaryDecoder';

interface PaneDaemonRequestFrame {
  type: 'request';
  id: number;
  channel: string;
  args: unknown[];
}

interface PaneDaemonSuccessResponseFrame {
  type: 'response';
  id: number;
  ok: true;
  result?: JsonValue;
}

interface PaneDaemonErrorResponseFrame {
  type: 'response';
  id: number;
  ok: false;
  error: {
    message: string;
    code?: string;
  };
}

interface PaneDaemonEventFrame {
  type: 'event';
  channel: string;
  args: unknown[];
}

type PaneDaemonFrame =
  | PaneDaemonRequestFrame
  | PaneDaemonSuccessResponseFrame
  | PaneDaemonErrorResponseFrame
  | PaneDaemonEventFrame;

export interface PaneDaemonEndpoint {
  transport: 'pipe' | 'unix';
  path: string;
}

interface InvokeOptions {
  paneDir?: string;
  timeoutMs?: number;
  eventInclude?: string[] | null;
}

interface TimeoutReference {
  current?: ReturnType<typeof setTimeout>;
}

type InvokeOutcome<Value> = { result: Value } | { error: Error };

const FRAME_DELIMITER = '\n';
const UNIX_SOCKET_BASE_DIRECTORY = '/tmp';
const DAEMON_SOCKET_FILENAME = 'daemon.sock';
const DEFAULT_TIMEOUT_MS = 130_000;

const paneDaemonFrameSchema: BoundarySchema<PaneDaemonFrame> = boundary.union(
  boundary.object({
    type: boundary.literal('request'),
    id: boundary.number,
    channel: boundary.string,
    args: boundary.array(boundary.json),
  }),
  boundary.object({
    type: boundary.literal('response'),
    id: boundary.number,
    ok: boundary.literal(true),
    result: boundary.optional(boundary.json),
  }),
  boundary.object({
    type: boundary.literal('response'),
    id: boundary.number,
    ok: boundary.literal(false),
    error: boundary.object({
      message: boundary.string,
      code: boundary.optional(boundary.string),
    }),
  }),
  boundary.object({
    type: boundary.literal('event'),
    channel: boundary.string,
    args: boundary.array(boundary.json),
  }),
);

export class PaneDaemonClientError extends Error {
  constructor(
    message: string,
    readonly code?: string,
  ) {
    super(message);
    this.name = 'PaneDaemonClientError';
  }
}

export function resolvePaneDirectory(paneDir?: string): string {
  return paneDir ?? process.env.PANE_DIR ?? process.env.FOOZOL_DIR ?? path.join(os.homedir(), '.pane');
}

export function getPaneDaemonEndpoint(appDirectory: string, platform: NodeJS.Platform = process.platform): PaneDaemonEndpoint {
  const resolvedAppDirectory = resolveAppDirectory(appDirectory, platform);

  if (platform === 'win32') {
    return {
      transport: 'pipe',
      path: getWindowsPipeName(resolvedAppDirectory),
    };
  }

  return {
    transport: 'unix',
    path: path.posix.join(getUnixSocketDirectoryName(resolvedAppDirectory), DAEMON_SOCKET_FILENAME),
  };
}

export async function invokeDaemon<T>(
  channel: string,
  args: unknown[] = [],
  resultSchema: BoundarySchema<T>,
  options: InvokeOptions = {},
): Promise<T> {
  const endpoint = getPaneDaemonEndpoint(resolvePaneDirectory(options.paneDir));
  const request: PaneDaemonRequestFrame = {
    type: 'request',
    id: 1,
    channel,
    args,
  };
  const eventFilterRequest: PaneDaemonRequestFrame = {
    type: 'request', id: 0, channel: 'daemon:events', args: [{ include: options.eventInclude ?? [] }],
  };

  return new Promise<T>((resolve, reject) => {
    const socket = net.createConnection(endpoint.path);
    const decoder = new PaneDaemonFrameDecoder();
    let settled = false;
    const timeoutRef: TimeoutReference = {};

    const settle = (outcome: InvokeOutcome<T>) => {
      if (settled) {
        return;
      }
      settled = true;
      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current);
      }
      socket.removeAllListeners();
      if (!socket.destroyed) {
        socket.destroy();
      }
      if ('error' in outcome) {
        reject(outcome.error);
        return;
      }
      resolve(outcome.result);
    };

    timeoutRef.current = setTimeout(() => {
      settle({ error: new PaneDaemonClientError(`Timed out waiting for Pane daemon response on ${endpoint.path}`, 'ERR_RUNPANE_DAEMON_TIMEOUT') });
    }, options.timeoutMs ?? DEFAULT_TIMEOUT_MS);

    socket.once('connect', () => {
      socket.write(encodePaneDaemonFrame(eventFilterRequest));
      socket.write(encodePaneDaemonFrame(request));
    });

    socket.on('data', (chunk) => {
      try {
        const frames = decoder.push(chunk);
        for (const frame of frames) {
          if (frame.type !== 'response' || frame.id !== request.id) {
            continue;
          }
          if (frame.ok) {
            settle({ result: decodeBoundary(frame.result, resultSchema) });
            return;
          }
          settle({ error: new PaneDaemonClientError(frame.error.message, frame.error.code) });
          return;
        }
      } catch (error) {
        settle({ error: error instanceof Error ? error : new Error(String(error)) });
      }
    });

    socket.once('error', (error: NodeJS.ErrnoException) => {
      const code = error.code ?? 'ERR_RUNPANE_DAEMON_CONNECT_FAILED';
      settle({ error: new PaneDaemonClientError(`Could not connect to Pane daemon at ${endpoint.path}: ${error.message}`, code) });
    });

    socket.once('close', () => {
      if (!settled) {
        settle({ error: new PaneDaemonClientError(`Pane daemon closed the connection before responding at ${endpoint.path}`, 'ERR_RUNPANE_DAEMON_CLOSED') });
      }
    });
  });
}

function resolveAppDirectory(appDirectory: string, platform: NodeJS.Platform): string {
  if (platform === 'win32') {
    return path.win32.resolve(appDirectory);
  }
  return path.posix.resolve(appDirectory);
}

function getWindowsPipeName(appDirectory: string): string {
  const hash = createHash('sha256')
    .update(appDirectory.toLowerCase())
    .digest('hex')
    .slice(0, 16);

  return `\\\\.\\pipe\\pane-daemon-${hash}`;
}

function getUnixSocketDirectoryName(appDirectory: string): string {
  const hash = createHash('sha256')
    .update(appDirectory)
    .digest('hex')
    .slice(0, 16);
  const uidSuffix = process.getuid ? `-${process.getuid()}` : '';

  return path.posix.join(UNIX_SOCKET_BASE_DIRECTORY, `pane-daemon${uidSuffix}-${hash}`);
}

function encodePaneDaemonFrame(frame: PaneDaemonFrame): string {
  return `${JSON.stringify(frame)}${FRAME_DELIMITER}`;
}

class PaneDaemonFrameDecoder {
  private buffer = '';
  private decoder = new StringDecoder('utf8');

  push(chunk: string | Buffer): PaneDaemonFrame[] {
    this.buffer += Buffer.isBuffer(chunk) ? this.decoder.write(chunk) : chunk;

    const frames: PaneDaemonFrame[] = [];
    let delimiterIndex = this.buffer.indexOf(FRAME_DELIMITER);

    while (delimiterIndex !== -1) {
      const rawFrame = this.buffer.slice(0, delimiterIndex);
      this.buffer = this.buffer.slice(delimiterIndex + FRAME_DELIMITER.length);

      if (rawFrame.trim().length > 0) {
        frames.push(parseFrame(rawFrame));
      }

      delimiterIndex = this.buffer.indexOf(FRAME_DELIMITER);
    }

    return frames;
  }
}

function parseFrame(rawFrame: string): PaneDaemonFrame {
  return decodeBoundary(JSON.parse(rawFrame), paneDaemonFrameSchema);
}

File: packages/runpane-py/src/runpane/daemon_client.py
from __future__ import annotations

import hashlib
import json
import ntpath
import os
import posixpath
import socket
import sys
from typing import Any, Dict, List, Optional

FRAME_DELIMITER = b"\n"
UNIX_SOCKET_BASE_DIRECTORY = "/tmp"
DAEMON_SOCKET_FILENAME = "daemon.sock"
DEFAULT_TIMEOUT_MS = 130_000


class PaneDaemonClientError(RuntimeError):
    def __init__(self, message: str, code: Optional[str] = None) -> None:
        super().__init__(message)
        self.code = code


def resolve_pane_directory(pane_dir: Optional[str] = None) -> str:
    return pane_dir or os.environ.get("PANE_DIR") or os.environ.get("FOOZOL_DIR") or os.path.join(os.path.expanduser("~"), ".pane")


def get_pane_daemon_endpoint(app_directory: str, platform: str = sys.platform) -> Dict[str, str]:
    resolved_app_directory = resolve_app_directory(app_directory, platform)
    if platform.startswith("win"):
        return {
            "transport": "pipe",
            "path": get_windows_pipe_name(resolved_app_directory),
        }

    return {
        "transport": "unix",
        "path": posixpath.join(get_unix_socket_directory_name(resolved_app_directory), DAEMON_SOCKET_FILENAME),
    }


def invoke_daemon(
    channel: str,
    args: Optional[List[Any]] = None,
    pane_dir: Optional[str] = None,
    timeout_ms: Optional[float] = None,
    event_include: Optional[List[str]] = None,
) -> Any:
    endpoint = get_pane_daemon_endpoint(resolve_pane_directory(pane_dir))
    request = {
        "type": "request",
        "id": 1,
        "channel": channel,
        "args": args or [],
    }
    encoded = encode_frame({
        "type": "request",
        "id": 0,
        "channel": "daemon:events",
        "args": [{"include": event_include or []}],
    })
    encoded += encode_frame(request)

    if endpoint["transport"] == "pipe":
        return invoke_windows_pipe(endpoint["path"], encoded, timeout_ms or DEFAULT_TIMEOUT_MS)
    return invoke_unix_socket(endpoint["path"], encoded, timeout_ms or DEFAULT_TIMEOUT_MS)


def invoke_unix_socket(socket_path: str, encoded_request: bytes, timeout_ms: float) -> Any:
    decoder = PaneDaemonFrameDecoder()
    with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as client:
        client.settimeout(timeout_ms / 1000)
        try:
            client.connect(socket_path)
        except OSError as error:
            raise PaneDaemonClientError(
                f"Could not connect to Pane daemon at {socket_path}: {error}",
                "ERR_RUNPANE_DAEMON_CONNECT_FAILED",
            ) from error

        client.sendall(encoded_request)
        while True:
            chunk = client.recv(65536)
            if not chunk:
                raise PaneDaemonClientError(
                    f"Pane daemon closed the connection before responding at {socket_path}",
                    "ERR_RUNPANE_DAEMON_CLOSED",
                )
            response = first_matching_response(decoder.push(chunk))
            if response is not None:
                return response


def invoke_windows_pipe(pipe_path: str, encoded_request: bytes, timeout_ms: float) -> Any:
    decoder = PaneDaemonFrameDecoder()
    try:
        with open(pipe_path, "r+b", buffering=0) as pipe:
            pipe.write(encoded_request)
            while True:
                chunk = pipe.read(65536)
                if not chunk:
                    raise PaneDaemonClientError(
                        f"Pane daemon closed the connection before responding at {pipe_path}",
                        "ERR_RUNPANE_DAEMON_CLOSED",
                    )
                response = first_matching_response(decoder.push(chunk))
                if response is not None:
                    return response
    except OSError as error:
        raise PaneDaemonClientError(
            f"Could not connect to Pane daemon at {pipe_path}: {error}",
            "ERR_RUNPANE_DAEMON_CONNECT_FAILED",
        ) from error


def first_matching_response(frames: List[Dict[str, Any]]) -> Optional[Any]:
    for frame in frames:
        if frame.get("type") != "response" or frame.get("id") != 1:
            continue
        if frame.get("ok") is True:
            return frame.get("result")
        error = frame.get("error") or {}
        raise PaneDaemonClientError(error.get("message", "Pane daemon request failed"), error.get("code"))
    return None


def resolve_app_directory(app_directory: str, platform: str) -> str:
    if platform.startswith("win"):
        return ntpath.abspath(app_directory)
    return posixpath.abspath(app_directory)


def get_windows_pipe_name(app_directory: str) -> str:
    digest = hashlib.sha256(app_directory.lower().encode("utf-8")).hexdigest()[:16]
    return "\\\\.\\pipe\\pane-daemon-" + digest


def get_unix_socket_directory_name(app_directory: str) -> str:
    digest = hashlib.sha256(app_directory.encode("utf-8")).hexdigest()[:16]
    uid_suffix = f"-{os.getuid()}" if hasattr(os, "getuid") else ""
    return posixpath.join(UNIX_SOCKET_BASE_DIRECTORY, f"pane-daemon{uid_suffix}-{digest}")


def encode_frame(frame: Dict[str, Any]) -> bytes:
    return json.dumps(frame, separators=(",", ":")).encode("utf-8") + FRAME_DELIMITER


class PaneDaemonFrameDecoder:
    def __init__(self) -> None:
        self.buffer = b""

    def push(self, chunk: bytes) -> List[Dict[str, Any]]:
        self.buffer += chunk
        frames: List[Dict[str, Any]] = []
        while FRAME_DELIMITER in self.buffer:
            raw_frame, self.buffer = self.buffer.split(FRAME_DELIMITER, 1)
            if raw_frame.strip():
                frames.append(json.loads(raw_frame.decode("utf-8")))
        return frames
